### ISSUE TITLE
Randomly missing audio and video

### ISSUE THREAD

**Issue by mcorner** [twilio/twilio-video-ios#36]
### Description

On a relatively small % of calls (P2P) we do each day we are missing either audio or video from a participant.   All of our calls are two participants, one is a client using iOS or Android, and our employee using the latest version of Chrome.

On the desktop side we see the remote tracks get subscribed.  Then we wait for the 'started' event, which also occurs.  But then we get no audio from our client or sometimes no video.  In most cases they will restart the call with the same result.  In one case we saw a reboot fix the issue.

We record the calls from the desktop side and there the audio/video is missing as well.  So our best guess is that this is a mobile client side problem.  The client is sending the streams, but they don't contain audio.

This happens on Android and iOS clients, though we probably see more on iOS.  Since we are talking about a broad range of issues, we would be happy to tackle just one: iOS audio.

We do not provide a mute button in our app so we are sure it isn't something disabling the audio track.

We are looking for ways to diagnose the cause and/or additional logging we can do.   Is it possible there is another mute somewhere?  Or something obvious we are missing?  Is #34 possibly related?  We don't unpublish tracks during a call though.

### Steps to Reproduce

Unfortunately we are unable to reproduce it locally and it only happens on < 5% of calls.

#### Code

Our iOS code is very similar to the quickstart code.  There isn't much to the audio portions,.

#### Expected Behavior

All tracks to work reliably.

#### Actual Behavior

Audio or video track starts but produces no data.

#### Reproduces How Often

<5% of calls.

#### Logs

We don't have any useful logs, but are looking for guidance on what to collect.

### Versions

ios:2.6.0
android: 2.0.1
js library: 2.0.0-beta4 (Always Chrome)

#### Video iOS SDK

See above.

#### Xcode

10.0.1

#### iOS Version

Various.  Doesn't seem to matter.

#### iOS Device

Various.  Doesn't seem to matter.

**Comment by ceaglest** [comment]
Hi @mcorner,

Thanks for the detailed report.

> We are looking for ways to diagnose the cause and/or additional logging we can do.

If you've got insights reporting enabled, then I think the best place to start would be if you could share Room SID(s) when the problem occurred. Let me think on this, but the level of device logging that would be helpful is probably not practical if you can't reproduce the issue locally.

> On the desktop side we see the remote tracks get subscribed. Then we wait for the 'started' event, which also occurs. But then we get no audio from our client or sometimes no video.

Okay, this means that ICE / DTLS layers are established and RTP is flowing. I'd be interested to see if bytes sent keeps increasing after this or not. If the audio device were not producing any audio samples I don't think any more bytes would be sent. If it was producing silence then that should be visible by the audio level stats.

Best,
Chris

**Comment by mcorner** [comment]
@ceaglest Thanks for the help.

I might have one clue on the audio problems at least.  One user rebooted and it worked, the other turned off bluetooth and it worked. So....I am wondering if these users' phones are connecting to their cars etc.  I would love to retain the ability for them to use bluetooth headsets, but it may also lead to user confusion.

On iOS can we:
a) Find out on iOS if it is using bluetooth and perhaps show an indicator?
b) turn off bluetooth entirely.

I can give you two examples to start with (we have more).  This is when the session ended (UTC) and the Room SID.  

Example 1:
iPhone XR (os 12.1.2).
The first two didn't have audio, the user rebooted and the the last one worked.

2019-02-06 02:32:56.808+00 | RM23a7488d3fdbffdd55b6816507574546
2019-02-06 02:34:22.489+00 | RMd29370f22693834818ec18a2c6710d22
2019-02-06 03:19:41.916+00 | RM7733e62113517372f7fd0c121253a833

Example 2:
No audio iPhone X (os 12.1).  In this case they reported turning off bluetooth and it worked :
2019-02-11 16:31:00.264+00 | RM7fbcb979e6e6d1ddf5e47e367fd1152c
2019-02-11 16:31:56.061+00 | RMbbbefa5cb92a0b5b9c2d2a2cd89afc32
2019-02-11 16:33:08.394+00 | RMc608ee3d81d402af60dd56af3cef757a
2019-02-11 16:34:10.146+00 | RM2386d20bde6d55a4de1a6c653c3a31af
2019-02-11 16:36:38.685+00 | RMef23fbb94e3be73c8b5705770146eeca
2019-02-11 17:09:41.369+00 | RM0e2d54b205653bfdcc0cc29c621a88af

**Comment by ceaglest** [comment]
Hi @mcorner,

> On iOS can we:
a) Find out on iOS if it is using bluetooth and perhaps show an indicator?

Yes, you could find this information out from `AVAudioSession.currentRoute`. Typically, the input and output are the same, but you want to dig down to the `AVAudioSessionPortDescription.portType` to be sure.

If the user swipes up they should be able to see the active audio route in control center, and a bluetooth icon in the status bar if a device is paired. Also you can add an [AVRoutePickerView](https://developer.apple.com/documentation/avkit/avroutepickerview) so that they can see the available routes from within your app (I have not tried this personally).

> b) turn off bluetooth entirely.

Yes, this is possible as well. If you are using `TVIDefaultAudioDevice` take a look at `kDefaultAVAudioSessionConfigurationBlock` and override [TVIDefaultAudioDevice.block](https://twilio.github.io/twilio-video-ios/docs/latest/Classes/TVIDefaultAudioDevice.html#//api/name/block) with the behavior that you want. We use AVAudioSessionCategoryOptionAllowBluetooth by default, but you don't need to set this if you don't want to.

> I can give you two examples to start with (we have more).

Thanks for these, sorry I didn't get to debug them today but I can have a look tomorrow.

Best,
Chris

**Comment by mcorner** [comment]
In case it helps someone else, here is what we added:

```
[[NSNotificationCenter` defaultCenter] addObserver:self
                                             selector:@selector(audioSessionRouteChanged:)
                                                 name:AVAudioSessionRouteChangeNotification
                                               object:nil];

- (void)audioSessionRouteChanged:(NSNotification *)notification {
  NSDictionary* info = notification.userInfo;

   AVAudioSessionRouteChangeReason reason = [info[AVAudioSessionRouteChangeReasonKey] unsignedIntValue];
   NSString *reasonDescription;
   switch (reason) {
     case AVAudioSessionRouteChangeReasonUnknown: {reasonDescription = @"AVAudioSessionRouteChangeReasonUnknown";} break;
     case AVAudioSessionRouteChangeReasonNewDeviceAvailable: {reasonDescription = @"AVAudioSessionRouteChangeReasonNewDeviceAvailable";} break;
     case AVAudioSessionRouteChangeReasonOldDeviceUnavailable: {reasonDescription = @"AVAudioSessionRouteChangeReasonOldDeviceUnavailable";} break;
     case AVAudioSessionRouteChangeReasonCategoryChange: {reasonDescription = @"AVAudioSessionRouteChangeReasonCategoryChange";} break;
     case AVAudioSessionRouteChangeReasonOverride: {reasonDescription = @"AVAudioSessionRouteChangeReasonOverride";} break;
     case AVAudioSessionRouteChangeReasonWakeFromSleep: {reasonDescription = @"AVAudioSessionRouteChangeReasonWakeFromSleep";} break;
     case AVAudioSessionRouteChangeReasonNoSuitableRouteForCategory: {reasonDescription = @"AVAudioSessionRouteChangeReasonNoSuitableRouteForCategory";} break;
     case AVAudioSessionRouteChangeReasonRouteConfigurationChange: {reasonDescription = @"AVAudioSessionRouteChangeReasonRouteConfigurationChange";} break;
     default: reasonDescription = @"Unknown";
   }

    AVAudioSessionRouteDescription *previousRoute = info[AVAudioSessionRouteChangePreviousRouteKey];
    NSString *previousInput = @"Unknown";
    if (previousRoute && [previousRoute.inputs count] > 0){
      previousInput = previousRoute.inputs[0].portType;
    }
    NSString *previousOutput = @"Unknown";
    if (previousRoute && [previousRoute.outputs count] > 0){
      previousOutput = previousRoute.outputs[0].portType;
    }

    AVAudioSessionRouteDescription *currentRoute = [AVAudioSession sharedInstance].currentRoute;
    NSString *currentInput = @"Unknown";
    if (currentRoute && [currentRoute.inputs count] > 0){
      currentInput = currentRoute.inputs[0].portType;
    }
    NSString *currentOutput = @"Unknown";
    if (currentRoute && [currentRoute.outputs count] > 0){
      currentOutput = currentRoute.outputs[0].portType;
    }

    if (![currentOutput isEqualToString:previousOutput] || ![currentInput isEqualToString:previousInput]){
      [self log:[NSString stringWithFormat:@"Audio route changed, reason = %@: previous = i:%@ o:%@ current = i:%@ o:%@", reasonDescription, previousInput, previousOutput, currentInput, currentOutput]];
    }
}
```

**Comment by ceaglest** [comment]
Hi @mcorner,

I looked at the Room SIDs that you provided (once again sorry for the delay). 

> #### iOS Version
> Various. Doesn't seem to matter.

I noticed that the iOS Client is using version 2.0.1. in each Room. We've made a few audio related bug fixes since then, so do you know if your customers are still experiencing no audio when using newer releases (like 2.7.0)?

> Example 1:
> iPhone XR (os 12.1.2).
> 2019-02-06 02:32:56.808+00 | RM23a7488d3fdbffdd55b6816507574546

Video looks good for the iOS Client, but very few audio bytes are sent (see MTb76179d0b93e2e2e09b7a83e95802015).

[RM23a7488d3fdbffdd55b6816507574546-track_infos.txt](https://github.com/twilio/twilio-video-ios/files/2916246/RM23a7488d3fdbffdd55b6816507574546-track_infos.txt)

> 2019-02-06 02:34:22.489+00 | RMd29370f22693834818ec18a2c6710d22

The iOS Client never joined this Room but the Chrome Client did. Since its Peer-to-Peer no media was exchanged.

> 2019-02-06 03:19:41.916+00 | RM7733e62113517372f7fd0c121253a833

There are a few reports missing for the Chrome Client, but I can see that the iOS Client is sending audio bytes (MTacdd99cf85854f153d5a3f2948074f15).

[RM7733e62113517372f7fd0c121253a833-track_infos.txt](https://github.com/twilio/twilio-video-ios/files/2916270/RM7733e62113517372f7fd0c121253a833-track_infos.txt)

> Example 2:
No audio iPhone X (os 12.1). In this case they reported turning off bluetooth and it worked :
2019-02-11 16:31:00.264+00 | RM7fbcb979e6e6d1ddf5e47e367fd1152c

I agree, I don't see any audio bytes being sent or received by the iOS device (running 2.0.1, MT3a5ae9a17a290f1b84a1a88152c7f786). Chrome tries to rejoin. If no bytes are being sent this likely means that the bluetooth audio device wasn't producing any output.

[RM7fbcb979e6e6d1ddf5e47e367fd1152c-track_infos.txt](https://github.com/twilio/twilio-video-ios/files/2916288/RM7fbcb979e6e6d1ddf5e47e367fd1152c-track_infos.txt)

> 2019-02-11 16:31:56.061+00 | RMbbbefa5cb92a0b5b9c2d2a2cd89afc32
> 2019-02-11 16:33:08.394+00 | RMc608ee3d81d402af60dd56af3cef757a
> 2019-02-11 16:34:10.146+00 | RM2386d20bde6d55a4de1a6c653c3a31af
> 2019-02-11 16:36:38.685+00 | RMef23fbb94e3be73c8b5705770146eeca

No audio bytes sent from iOS.

> 2019-02-11 17:09:41.369+00 | RM0e2d54b205653bfdcc0cc29c621a88af

Finally, bytes are sent from the iOS Client (MT7cef7ecc471af70795464ba68f156e61).

[RM0e2d54b205653bfdcc0cc29c621a88af-track_infos.txt](https://github.com/twilio/twilio-video-ios/files/2916308/RM0e2d54b205653bfdcc0cc29c621a88af-track_infos.txt)

In summary, I think you are on the right path with letting your customers know about route changes so that they can better tell if Bluetooth is being used. Also, you should consider disabling bluetooth routes if it doesn't make sense for your use case (or adding an app setting to enable the routes).

Regards,
Chris

**Comment by mcorner** [comment]
@ceaglest Awesome analysis.  Yes, we have users on old versions of our app.  We are planning to persuade them to upgrade in our next revision.  We will only report back on problems on most recent SDKs.  Instead of showing them an indication or disabling bluetooth, we are going to just start with logging.  Then we can do a post-hoc analysis to see if Bluetooth is indeed the issue.

I will report back, thanks again.

**Comment by mcorner** [comment]
We aren't really seeing this specific problem after upgrading.  It is hard for us to always track these, but I will close this for now.

**Comment by ceaglest** [comment]
Thanks for letting us know @mcorner.
